### PR TITLE
Technical Test - Nestor Angelo A. Gimenez

### PR DESCRIPTION
**Pull request by ayofish**
Hi,

This application is built using nodejs with a simple angular js front-end and with mongodb as the persistence layer. 

To run the application you will need to install "mongodb" and nodejs on your computer. then just open your terminal and go to the folder and type in "npm install && bower install" this will install the packages for the client side and the server side. Once installation is complete please type in "npm start" and open your favorite browser then type in localhost:3000. This will open the application.

The design pattern I used is mvc in the front-end and dependency injection. AngualrJs uses these two concepts very well and makes development faster. Dependency injection was used by angular when defining the modules that will be used similar to Asynchronous module definition like requirejs.

For the backend I used the express framework from node. This is the most used framework by node developers since it provides the basic functionalities of routing and content distribution. For the api I used the concept of REST which is representational state transfer by using endpoints that are differentiated by the server side via the http request methods or verbs such as post, put, get, delete. This is the best way in communicating with a server since in normal production cases authentication is done on every request and there is no state held on the server side. For the architecture I leveraged the module pattern which comes with nodejs by default. This enables me to encapsulate code in each js module.

I also used mongoose to handle the mongodb connection and data layer of the application since using an object document mapper such as this is good for data consistency specially since mongodb has inconsistent data. It just takes a  lot of headaches away.

Other libraries could have been used for the server side such as koa or restify. I chose express framework since it is the one that I am most familiar with and because it holds a lot of functionality without really modifying how nodejs is meant to work.

For the front-end I chose angualr js since it is the one that I have the most fun with. I could have used backbone or just vanilla javascript or jquery with some plugins but I felt nodejs would be the best fit since it's data binding feature is so powerful as you will see once you run my application. Plus angularjs was designed originally for fast prototyping or development and it holds true to that up until this day.

Major challenges that I encountered while developing this application is the time. I work a full time job as a developer and this week was a hectic week at work. Another challenge is that my second monitor just broke down so I developed this using just one monitor. Alt+tab was a bit challenging and tedious.

If I had more time I would have added a dashboard for the user and authentication. The dashboard would have had widgets that contain the notes and the todos in calendar or just a list at the side for the user's quick viewing.

Thanks and Regards,
Nestor Angelo Gimenez

### CODE REVIEW

**Comment by zakame** [comment]
Seems to lack build instructions? `npm install && bower install`

File: routes.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ */
+
+
+var path = require('path');
+
+module.exports = function(app) {
+  // Insert routes below
+  app.use('/api/todos', require('./api/todos'));
+  app.use('/api/notes', require('./api/notes'));
+
```
Comment (zakame) [inline routes.js]: No route for `/api` itself?  At least it should handle and respond with pointers to `/api/notes` and `/api/todos`

**Comment by zakame** [comment]
Also, no way to convert between notes and todos, per spec:

```
The app should make it easy to switch a note to a TODO entry, and vice versa.
```

**Comment by ayofish** [comment]
@zakame  in my pull request I stated: "To run the application you will need to install "mongodb" and nodejs on your computer. then just open your terminal and go to the folder and type in "npm install && bower install" this will install the packages for the client side and the server side. Once installation is complete please type in "npm start" and open your favorite browser then type in localhost:3000. This will open the application." after the bower install and npm install just do an npm start and it will run the applicaiton.

**Comment by ayofish** [comment]
@zakame Regarding no route for /api, yup I did not put a specific action attached to /api I did not find it necessary.

**Comment by ayofish** [comment]
@zakame I missed this part of the instructions:
"The app should make it easy to switch a note to a TODO entry, and vice versa."

I can make a few changes and commit what's missing.

**Comment by zakame** [comment]
@ayofish yep, kindly update to follow spec :)  Thanks!

**Comment by ayofish** [comment]
@zakame I made the necessary changes for converting a Todo Item to a note and vice versa.

**Pull request closed without merging**